File: supabase/functions/_shared/popia-compliance.ts
/**
 * POPIA (Protection of Personal Information Act) Compliance Utilities
 *
 * This module provides PII detection, redaction, and compliance tracking
 * for data being sent to external AI APIs.
 *
 * Key POPIA principles enforced:
 * 1. Data minimization - only send necessary data
 * 2. Purpose limitation - data used only for stated purpose
 * 3. Security safeguards - redaction of sensitive information
 * 4. Accountability - audit logging of what was sent
 */

// South African specific PII patterns
const SA_ID_PATTERN = /\b\d{13}\b/g;  // SA ID number (13 digits)
const SA_PASSPORT_PATTERN = /\b[A-Z]\d{8}\b/gi;  // SA Passport
const SA_PHONE_PATTERN = /\b(?:\+27|0)[1-9]\d{8,9}\b/g;  // SA phone numbers
const SA_BANK_ACCOUNT_PATTERN = /\b\d{10,16}\b/g;  // Bank account numbers (contextual)

// General PII patterns
const EMAIL_PATTERN = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g;
const CREDIT_CARD_PATTERN = /\b(?:\d{4}[-\s]?){3}\d{4}\b/g;
const IP_ADDRESS_PATTERN = /\b(?:\d{1,3}\.){3}\d{1,3}\b/g;
const DATE_OF_BIRTH_PATTERN = /\b(?:DOB|Date of Birth|Born|Birth Date)[:\s]*\d{1,2}[-\/]\d{1,2}[-\/]\d{2,4}\b/gi;

// Name patterns (contextual - used with keywords)
const NAME_KEYWORDS = ['name:', 'naam:', 'full name:', 'surname:', 'first name:', 'last name:', 'applicant:', 'client:', 'customer:', 'employee:'];
const ADDRESS_KEYWORDS = ['address:', 'adres:', 'street:', 'road:', 'avenue:', 'postal:', 'physical address:', 'residential:'];

export interface PIIDetectionResult {
  hasPII: boolean;
  detectedTypes: string[];
  matches: PIIMatch[];
  riskLevel: 'none' | 'low' | 'medium' | 'high';
  recommendations: string[];
}

export interface PIIMatch {
  type: string;
  pattern: string;
  count: number;
  sample: string;  // Partially masked sample for logging
}

export interface RedactionOptions {
  redactEmails: boolean;
  redactPhones: boolean;
  redactIDNumbers: boolean;
  redactBankAccounts: boolean;
  redactCreditCards: boolean;
  redactAddresses: boolean;
  redactNames: boolean;
  customPatterns?: RegExp[];
  replacementChar?: string;
}

export const DEFAULT_REDACTION_OPTIONS: RedactionOptions = {
  redactEmails: true,
  redactPhones: true,
  redactIDNumbers: true,
  redactBankAccounts: false,  // Too many false positives with numbers
  redactCreditCards: true,
  redactAddresses: false,  // Context-dependent
  redactNames: false,  // Context-dependent
  replacementChar: '*',
};

/**
 * Detect PII in text content
 */
export function detectPII(text: string): PIIDetectionResult {
  const matches: PIIMatch[] = [];
  const detectedTypes: string[] = [];

  // Check SA ID Numbers
  const idMatches = text.match(SA_ID_PATTERN);
  if (idMatches && idMatches.length > 0) {
    // Validate SA ID format (basic Luhn check could be added)
    const validIds = idMatches.filter(id => isValidSAID(id));
    if (validIds.length > 0) {
      matches.push({
        type: 'SA ID Number',
        pattern: 'XXXXXX****XXX',
        count: validIds.length,
        sample: maskPII(validIds[0], 'id'),
      });
      detectedTypes.push('SA ID Number');
    }
  }

  // Check Email Addresses
  const emailMatches = text.match(EMAIL_PATTERN);
  if (emailMatches && emailMatches.length > 0) {
    matches.push({
      type: 'Email Address',
      pattern: 'x***@***.***',
      count: emailMatches.length,
      sample: maskPII(emailMatches[0], 'email'),
    });
    detectedTypes.push('Email Address');
  }

  // Check Phone Numbers
  const phoneMatches = text.match(SA_PHONE_PATTERN);
  if (phoneMatches && phoneMatches.length > 0) {
    matches.push({
      type: 'Phone Number',
      pattern: '+27*****XXXX',
      count: phoneMatches.length,
      sample: maskPII(phoneMatches[0], 'phone'),
    });
    detectedTypes.push('Phone Number');
  }

  // Check Credit Cards
  const ccMatches = text.match(CREDIT_CARD_PATTERN);
  if (ccMatches && ccMatches.length > 0) {
    matches.push({
      type: 'Credit Card',
      pattern: '****-****-****-XXXX',
      count: ccMatches.length,
      sample: maskPII(ccMatches[0], 'credit_card'),
    });
    detectedTypes.push('Credit Card');
  }

  // Check for contextual PII (names, addresses)
  const lowerText = text.toLowerCase();
  const hasNameContext = NAME_KEYWORDS.some(kw => lowerText.includes(kw));
  const hasAddressContext = ADDRESS_KEYWORDS.some(kw => lowerText.includes(kw));

  if (hasNameContext) {
    detectedTypes.push('Potential Name Reference');
  }
  if (hasAddressContext) {
    detectedTypes.push('Potential Address Reference');
  }

  // Check Date of Birth
  const dobMatches = text.match(DATE_OF_BIRTH_PATTERN);
  if (dobMatches && dobMatches.length > 0) {
    matches.push({
      type: 'Date of Birth',
      pattern: 'DOB: **/**/****',
      count: dobMatches.length,
      sample: '[DATE OF BIRTH]',
    });
    detectedTypes.push('Date of Birth');
  }

  // Calculate risk level
  const riskLevel = calculateRiskLevel(detectedTypes, matches);

  // Generate recommendations
  const recommendations = generateRecommendations(detectedTypes, riskLevel);

  return {
    hasPII: detectedTypes.length > 0,
    detectedTypes,
    matches,
    riskLevel,
    recommendations,
  };
}

/**
 * Redact PII from text based on options
 */
export function redactPII(text: string, options: Partial<RedactionOptions> = {}): string {
  const opts = { ...DEFAULT_REDACTION_OPTIONS, ...options };
  let redacted = text;
  const char = opts.replacementChar || '*';

  if (opts.redactIDNumbers) {
    redacted = redacted.replace(SA_ID_PATTERN, (match) => {
      if (isValidSAID(match)) {
        return match.slice(0, 6) + char.repeat(4) + match.slice(10);
      }
      return match;
    });
  }

  if (opts.redactEmails) {
    redacted = redacted.replace(EMAIL_PATTERN, (match) => {
      const [local, domain] = match.split('@');
      return local[0] + char.repeat(Math.min(local.length - 1, 5)) + '@' + char.repeat(3) + '.' + domain.split('.').pop();
    });
  }

  if (opts.redactPhones) {
    redacted = redacted.replace(SA_PHONE_PATTERN, (match) => {
      return match.slice(0, 3) + char.repeat(match.length - 7) + match.slice(-4);
    });
  }

  if (opts.redactCreditCards) {
    redacted = redacted.replace(CREDIT_CARD_PATTERN, (match) => {
      const digits = match.replace(/[-\s]/g, '');
      return char.repeat(12) + digits.slice(-4);
    });
  }

  // Custom patterns
  if (opts.customPatterns) {
    for (const pattern of opts.customPatterns) {
      redacted = redacted.replace(pattern, char.repeat(8));
    }
  }

  return redacted;
}

/**
 * Create a compliance audit log entry
 */
export interface ComplianceAuditEntry {
  timestamp: string;
  action: string;
  dataType: string;
  piiDetected: PIIDetectionResult;
  redactionApplied: boolean;
  dataSentBytes: number;
  destinationService: string;
  purpose: string;
  userConsented: boolean;
}

export function createAuditEntry(
  action: string,
  dataType: string,
  piiResult: PIIDetectionResult,
  redactionApplied: boolean,
  dataSentBytes: number,
  destinationService: string,
  purpose: string,
): ComplianceAuditEntry {
  return {
    timestamp: new Date().toISOString(),
    action,
    dataType,
    piiDetected: piiResult,
    redactionApplied,
    dataSentBytes,
    destinationService,
    purpose,
    userConsented: true,  // Consent assumed through app usage
  };
}

/**
 * Generate a compliance summary for the user
 */
export function generateComplianceSummary(auditEntry: ComplianceAuditEntry): string {
  const lines: string[] = [];

  lines.push(`=== POPIA Compliance Summary ===`);
  lines.push(`Timestamp: ${auditEntry.timestamp}`);
  lines.push(`Action: ${auditEntry.action}`);
  lines.push(`Destination: ${auditEntry.destinationService}`);
  lines.push(`Purpose: ${auditEntry.purpose}`);
  lines.push(`Data Size: ${(auditEntry.dataSentBytes / 1024).toFixed(2)} KB`);

  if (auditEntry.piiDetected.hasPII) {
    lines.push(`\nPII Detected: YES`);
    lines.push(`Risk Level: ${auditEntry.piiDetected.riskLevel.toUpperCase()}`);
    lines.push(`Types Found: ${auditEntry.piiDetected.detectedTypes.join(', ')}`);
    lines.push(`Redaction Applied: ${auditEntry.redactionApplied ? 'YES' : 'NO'}`);

    if (!auditEntry.redactionApplied && auditEntry.piiDetected.riskLevel !== 'none') {
      lines.push(`\n⚠️ WARNING: PII was sent without redaction`);
    }
  } else {
    lines.push(`\nPII Detected: NO`);
  }

  return lines.join('\n');
}

// Helper functions

function isValidSAID(id: string): boolean {
  // Basic SA ID validation
  // Format: YYMMDD SSSS C A Z
  // YY = Year, MM = Month, DD = Day
  // SSSS = Sequence, C = Citizenship, A = Usually 8, Z = Checksum

  if (id.length !== 13) return false;

  const year = parseInt(id.slice(0, 2));
  const month = parseInt(id.slice(2, 4));
  const day = parseInt(id.slice(4, 6));

  // Basic date validation
  if (month < 1 || month > 12) return false;
  if (day < 1 || day > 31) return false;

  // Luhn algorithm check
  let sum = 0;
  for (let i = 0; i < 13; i++) {
    let digit = parseInt(id[i]);
    if (i % 2 === 1) {
      digit *= 2;
      if (digit > 9) digit -= 9;
    }
    sum += digit;
  }

  return sum % 10 === 0;
}

function maskPII(value: string, type: string): string {
  switch (type) {
    case 'id':
      return value.slice(0, 6) + '****' + value.slice(10);
    case 'email':
      const [local, domain] = value.split('@');
      return local[0] + '***@***.' + domain.split('.').pop();
    case 'phone':
      return value.slice(0, 3) + '*****' + value.slice(-4);
    case 'credit_card':
      return '****-****-****-' + value.replace(/[-\s]/g, '').slice(-4);
    default:
      return '***';
  }
}

function calculateRiskLevel(types: string[], matches: PIIMatch[]): 'none' | 'low' | 'medium' | 'high' {
  if (types.length === 0) return 'none';

  const highRiskTypes = ['SA ID Number', 'Credit Card', 'Bank Account'];
  const mediumRiskTypes = ['Email Address', 'Phone Number', 'Date of Birth'];

  const hasHighRisk = types.some(t => highRiskTypes.includes(t));
  const hasMediumRisk = types.some(t => mediumRiskTypes.includes(t));
  const totalMatches = matches.reduce((sum, m) => sum + m.count, 0);

  if (hasHighRisk || totalMatches > 10) return 'high';
  if (hasMediumRisk || totalMatches > 5) return 'medium';
  return 'low';
}

function generateRecommendations(types: string[], riskLevel: string): string[] {
  const recommendations: string[] = [];

  if (riskLevel === 'high') {
    recommendations.push('Consider enabling PII redaction before sending to external APIs');
    recommendations.push('Review document content for unnecessary personal information');
  }

  if (types.includes('SA ID Number')) {
    recommendations.push('SA ID numbers detected - ensure compliance with POPIA Section 26 (Special Personal Information)');
  }

  if (types.includes('Email Address') || types.includes('Phone Number')) {
    recommendations.push('Contact information detected - verify purpose limitation compliance');
  }

  if (riskLevel !== 'none') {
    recommendations.push('Consider using anonymized or aggregated data where possible');
    recommendations.push('Ensure data retention policies are followed');
  }

  return recommendations;
}

/**
 * Prepare text for API call with POPIA compliance
 * Returns the processed text and compliance metadata
 */
export interface PreparedDataResult {
  text: string;
  originalLength: number;
  processedLength: number;
  piiDetection: PIIDetectionResult;
  wasRedacted: boolean;
  auditEntry: ComplianceAuditEntry;
}

export function prepareForAPICall(
  text: string,
  options: {
    enableRedaction: boolean;
    redactionOptions?: Partial<RedactionOptions>;
    purpose: string;
    destinationService: string;
  }
): PreparedDataResult {
  const originalLength = text.length;

  // Detect PII
  const piiDetection = detectPII(text);

  // Apply redaction if enabled and PII detected
  let processedText = text;
  let wasRedacted = false;

  if (options.enableRedaction && piiDetection.hasPII) {
    processedText = redactPII(text, options.redactionOptions);
    wasRedacted = true;
  }

  // Create audit entry
  const auditEntry = createAuditEntry(
    'API_CALL',
    'document_text',
    piiDetection,
    wasRedacted,
    processedText.length,
    options.destinationService,
    options.purpose,
  );

  return {
    text: processedText,
    originalLength,
    processedLength: processedText.length,
    piiDetection,
    wasRedacted,
    auditEntry,
  };
}

File: backend-py/services/popia_compliance.py
"""POPIA (Protection of Personal Information Act) Compliance Utilities.

PII detection, redaction, and compliance tracking for data sent to external AI APIs.
"""

import re
from datetime import datetime, timezone

# South African specific PII patterns
SA_ID_PATTERN = re.compile(r"\b\d{13}\b")
SA_PASSPORT_PATTERN = re.compile(r"\b[A-Z]\d{8}\b", re.IGNORECASE)
SA_PHONE_PATTERN = re.compile(r"\b(?:\+27|0)[1-9]\d{8,9}\b")

# General PII patterns
EMAIL_PATTERN = re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b")
CREDIT_CARD_PATTERN = re.compile(r"\b(?:\d{4}[-\s]?){3}\d{4}\b")
DATE_OF_BIRTH_PATTERN = re.compile(
    r"\b(?:DOB|Date of Birth|Born|Birth Date)[:\s]*\d{1,2}[-/]\d{1,2}[-/]\d{2,4}\b",
    re.IGNORECASE,
)

NAME_KEYWORDS = [
    "name:", "naam:", "full name:", "surname:", "first name:",
    "last name:", "applicant:", "client:", "customer:", "employee:",
]
ADDRESS_KEYWORDS = [
    "address:", "adres:", "street:", "road:", "avenue:",
    "postal:", "physical address:", "residential:",
]

DEFAULT_REDACTION_OPTIONS = {
    "redactEmails": True,
    "redactPhones": True,
    "redactIDNumbers": True,
    "redactBankAccounts": False,
    "redactCreditCards": True,
    "redactAddresses": False,
    "redactNames": False,
    "replacementChar": "*",
}


def is_valid_sa_id(id_str: str) -> bool:
    """Basic SA ID validation with Luhn check."""
    if len(id_str) != 13:
        return False

    month = int(id_str[2:4])
    day = int(id_str[4:6])

    if month < 1 or month > 12:
        return False
    if day < 1 or day > 31:
        return False

    # Luhn algorithm check
    total = 0
    for i in range(13):
        digit = int(id_str[i])
        if i % 2 == 1:
            digit *= 2
            if digit > 9:
                digit -= 9
        total += digit

    return total % 10 == 0


def _mask_pii(value: str, pii_type: str) -> str:
    if pii_type == "id":
        return value[:6] + "****" + value[10:]
    elif pii_type == "email":
        parts = value.split("@")
        domain_parts = parts[1].split(".")
        return parts[0][0] + "***@***." + domain_parts[-1]
    elif pii_type == "phone":
        return value[:3] + "*****" + value[-4:]
    elif pii_type == "credit_card":
        digits = re.sub(r"[-\s]", "", value)
        return "****-****-****-" + digits[-4:]
    return "***"


def detect_pii(text: str) -> dict:
    """Detect PII in text content.

    Returns dict with: hasPII, detectedTypes, matches, riskLevel, recommendations.
    """
    matches = []
    detected_types = []

    # SA ID Numbers
    id_matches = SA_ID_PATTERN.findall(text)
    valid_ids = [m for m in id_matches if is_valid_sa_id(m)]
    if valid_ids:
        matches.append({
            "type": "SA ID Number",
            "pattern": "XXXXXX****XXX",
            "count": len(valid_ids),
            "sample": _mask_pii(valid_ids[0], "id"),
        })
        detected_types.append("SA ID Number")

    # Email Addresses
    email_matches = EMAIL_PATTERN.findall(text)
    if email_matches:
        matches.append({
            "type": "Email Address",
            "pattern": "x***@***.***",
            "count": len(email_matches),
            "sample": _mask_pii(email_matches[0], "email"),
        })
        detected_types.append("Email Address")

    # Phone Numbers
    phone_matches = SA_PHONE_PATTERN.findall(text)
    if phone_matches:
        matches.append({
            "type": "Phone Number",
            "pattern": "+27*****XXXX",
            "count": len(phone_matches),
            "sample": _mask_pii(phone_matches[0], "phone"),
        })
        detected_types.append("Phone Number")

    # Credit Cards
    cc_matches = CREDIT_CARD_PATTERN.findall(text)
    if cc_matches:
        matches.append({
            "type": "Credit Card",
            "pattern": "****-****-****-XXXX",
            "count": len(cc_matches),
            "sample": _mask_pii(cc_matches[0], "credit_card"),
        })
        detected_types.append("Credit Card")

    # Contextual PII
    lower_text = text.lower()
    if any(kw in lower_text for kw in NAME_KEYWORDS):
        detected_types.append("Potential Name Reference")
    if any(kw in lower_text for kw in ADDRESS_KEYWORDS):
        detected_types.append("Potential Address Reference")

    # Date of Birth
    dob_matches = DATE_OF_BIRTH_PATTERN.findall(text)
    if dob_matches:
        matches.append({
            "type": "Date of Birth",
            "pattern": "DOB: **/**/****",
            "count": len(dob_matches),
            "sample": "[DATE OF BIRTH]",
        })
        detected_types.append("Date of Birth")

    risk_level = _calculate_risk_level(detected_types, matches)
    recommendations = _generate_recommendations(detected_types, risk_level)

    return {
        "hasPII": len(detected_types) > 0,
        "detectedTypes": detected_types,
        "matches": matches,
        "riskLevel": risk_level,
        "recommendations": recommendations,
    }


def redact_pii(text: str, options: dict | None = None) -> str:
    """Redact PII from text based on options."""
    opts = {**DEFAULT_REDACTION_OPTIONS, **(options or {})}
    redacted = text
    char = opts.get("replacementChar", "*")

    if opts["redactIDNumbers"]:
        def _redact_id(m):
            val = m.group()
            if is_valid_sa_id(val):
                return val[:6] + char * 4 + val[10:]
            return val
        redacted = SA_ID_PATTERN.sub(_redact_id, redacted)

    if opts["redactEmails"]:
        def _redact_email(m):
            val = m.group()
            parts = val.split("@")
            local = parts[0]
            domain = parts[1]
            return local[0] + char * min(len(local) - 1, 5) + "@" + char * 3 + "." + domain.split(".")[-1]
        redacted = EMAIL_PATTERN.sub(_redact_email, redacted)

    if opts["redactPhones"]:
        def _redact_phone(m):
            val = m.group()
            return val[:3] + char * (len(val) - 7) + val[-4:]
        redacted = SA_PHONE_PATTERN.sub(_redact_phone, redacted)

    if opts["redactCreditCards"]:
        def _redact_cc(m):
            val = m.group()
            digits = re.sub(r"[-\s]", "", val)
            return char * 12 + digits[-4:]
        redacted = CREDIT_CARD_PATTERN.sub(_redact_cc, redacted)

    return redacted


def create_audit_entry(
    action: str,
    data_type: str,
    pii_result: dict,
    redaction_applied: bool,
    data_sent_bytes: int,
    destination_service: str,
    purpose: str,
) -> dict:
    return {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "action": action,
        "dataType": data_type,
        "piiDetected": pii_result,
        "redactionApplied": redaction_applied,
        "dataSentBytes": data_sent_bytes,
        "destinationService": destination_service,
        "purpose": purpose,
        "userConsented": True,
    }


def prepare_for_api_call(text: str, options: dict) -> dict:
    """Prepare text for API call with POPIA compliance. Returns processed text + metadata."""
    original_length = len(text)
    pii_detection = detect_pii(text)

    processed_text = text
    was_redacted = False

    if options.get("enableRedaction") and pii_detection["hasPII"]:
        processed_text = redact_pii(text, options.get("redactionOptions"))
        was_redacted = True

    audit_entry = create_audit_entry(
        "API_CALL", "document_text", pii_detection, was_redacted,
        len(processed_text), options.get("destinationService", ""), options.get("purpose", ""),
    )

    return {
        "text": processed_text,
        "originalLength": original_length,
        "processedLength": len(processed_text),
        "piiDetection": pii_detection,
        "wasRedacted": was_redacted,
        "auditEntry": audit_entry,
    }


def _calculate_risk_level(types: list[str], matches: list[dict]) -> str:
    if not types:
        return "none"

    high_risk = {"SA ID Number", "Credit Card", "Bank Account"}
    medium_risk = {"Email Address", "Phone Number", "Date of Birth"}

    has_high = any(t in high_risk for t in types)
    has_medium = any(t in medium_risk for t in types)
    total_matches = sum(m["count"] for m in matches)

    if has_high or total_matches > 10:
        return "high"
    if has_medium or total_matches > 5:
        return "medium"
    return "low"


def _generate_recommendations(types: list[str], risk_level: str) -> list[str]:
    recs: list[str] = []

    if risk_level == "high":
        recs.append("Consider enabling PII redaction before sending to external APIs")
        recs.append("Review document content for unnecessary personal information")

    if "SA ID Number" in types:
        recs.append("SA ID numbers detected - ensure compliance with POPIA Section 26 (Special Personal Information)")

    if "Email Address" in types or "Phone Number" in types:
        recs.append("Contact information detected - verify purpose limitation compliance")

    if risk_level != "none":
        recs.append("Consider using anonymized or aggregated data where possible")
        recs.append("Ensure data retention policies are followed")

    return recs
